feat(workflow): GroupOwner migration

iProgramStuff opened the pull request:
Adding a data model to support new features around "suggested assignees" or "group owners". 

We want to be able to have this data displayed in the issue search and searchable in a performant manner.

k-fish committed:
ref(ts): Convert linechart (again) (#22144)

Fixing revert #21823 which started failing in master despite PR passing checks (likely merge issues)

iProgramStuff committed:
Adding GroupOwner migration

iProgramStuff committed:
Fixing a typo

iProgramStuff committed:
Updating migration

iProgramStuff commented on src/sentry/migrations/0130_groupownermodel.py:
```diff
@@ -0,0 +1,51 @@
+    #   they can be monitored. Since data migrations will now hold a transaction open
+    #   this is even more important.
+    # - Adding columns to highly active tables, even ones that are NULL.
+    is_dangerous = False
```
Since this is a brand new table, it is okay for this to be False.

iProgramStuff commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,34 @@
+    owner_type = models.PositiveSmallIntegerField(null=False)
+    # ID of team/user, based on owner_type
+    # Wondering if this should just be team_id and user_id as nullable foreign keys and one is just not used per entry.
+    owner_id = BoundedBigIntegerField(null=False)
```
Not sure whether two foreign keys to user/team is better, or this.

iProgramStuff committed:
feat(workflow): Adding json schema for GroupInbox reason_details (#21882)

* Adding json schema for GroupInbox reason_details

* Slipping in SENTRY-JG9 fix

Zylphrex committed:
fix(releases): Explicitly filter transactions list by event type (#22160)

Without explicitly filtering on transactions, the open in discover button can
bring the user to some confusing queries. For example, other event types can
appear in the chart/meta because they're not being filtered out but is not
appearing in the table because the `p50()` column is causing them to not appear.

evanpurkhiser committed:
chore(js): Update framer-motion to v2.x (#22119)

scttcper committed:
perf: Collapse latestDeploys in project component (#22155)

dashed committed:
feat(dashboardsV2): Init Dashboards V2 (#22043)

Co-authored-by: Mark Story <mark@sentry.io>

iProgramStuff committed:
Updating model based on convo with Matt

iProgramStuff committed:
.

Zylphrex committed:
fix(releases): Remove sort by transaction from list and update copy (#22156)

Removed the option to sort by transaction name from the transactions list
dropdown and update the copy for the dropdown.

mattrobenolt committed:
fix(net): Gracefully handle invalid DNS names

If a hostname can't be IDNA encoded, it raises a UnicodeError, so we want to handle this the same way as a failed DNS lookup.

billyvg committed:
ref(ts): Refactor `<OrganizationApiKeys>` (#22158)

Note this is deprecated, but I want to remove reflexbox in the List component next

iProgramStuff commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+    project = FlexibleForeignKey("sentry.Project", null=False, db_constraint=False)
+    organization = FlexibleForeignKey("sentry.Organization", null=False, db_constraint=False)
+    owner_type = models.PositiveSmallIntegerField(null=False)
+    user = FlexibleForeignKey(settings.AUTH_USER_MODEL, related_name="sentry_owner_set", null=True)
+    team = FlexibleForeignKey("sentry.Team", related_name="sentry_owner_set", null=True)
```
Should these be marked as `db_constraint=False`?

wedamija replied:
I think these should be safe.

mattrobenolt committed:
perf: Remove a query in post_porcess_group

Bind on group.project pre-emptively since we already have it in memory.
This prevents a future query when hitting group.project.

Zylphrex committed:
fix(releases): Swap failure rate for failure count in releases (#22165)

Swap failure rate for failure count in the transactions list in releases. The
failure rate is susceptible to low counts, making it not very valuable. Failure
count, however, does not suffer from this.

billyvg committed:
ref(ts): Refactor `<OrganizationApiKeysList>` (#22163)

This also removes reflexbox and replaces with `<PanelTable>`

billyvg committed:
ref(ui): Rename `<StyledSettingsPageHeading>` (#22161)

Change from `Heading` -> `Header` to match file name. Also change default export to match file name. This will help LSP with auto importing.

iProgramStuff committed:
feat(workflow): Support `base` collapse and new endpoint to fetch stats (#21986)

davidenwang committed:
style(metric_alerts): Metric alert builder polish (#22140)

Some very small text/padding changes from feedback on the new metric alert builder form.

"Select the events you want to alert on" -> "Select events"
"All Environments" -> "All" ("Env:" -> "Environment:")
Increased size of filter bar and having it only appear on the same row as the other selectors for large monitors (not laptops)
Removed excess vertical whitespace in between steps 1 & 2 as well as between selectors
Total events updates correctly when changing time window (which can cause time period to change)
Additional polishing is done separately in #22133, #22130

billyvg committed:
build(babel): Remove un-needed babel plugins (#21701)

* Removes object spread plugin, this is included in `@babel/preset-env`
* Removes syntax dynamic import, which is supported by webpack (and `dynamic-import-node` for jest)

mattrobenolt commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+        super(GroupOwner, self).save(*args, **kwargs)
+
+    def owner_id(self):
+        if self.user:
```
```suggestion
        if self.user_id:
```

mattrobenolt commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+        if self.user:
+            return u"user:{}".format(self.user_id)
+
+        if self.team:
```
```suggestion
        if self.team_id:
```

mattrobenolt commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+    __core__ = False
+
+    group = FlexibleForeignKey("sentry.Group", unique=True, db_constraint=False)
+    project = FlexibleForeignKey("sentry.Project", null=False, db_constraint=False)
+    organization = FlexibleForeignKey("sentry.Organization", null=False, db_constraint=False)
+    owner_type = models.PositiveSmallIntegerField(null=False)
```
Should use `choices` here since it's directly from the GroupOwnerType enum.

wedamija replied:
Nit: No need to set `null=False`

mattrobenolt commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+    project = FlexibleForeignKey("sentry.Project", null=False, db_constraint=False)
+    organization = FlexibleForeignKey("sentry.Organization", null=False, db_constraint=False)
+    owner_type = models.PositiveSmallIntegerField(null=False)
+    user = FlexibleForeignKey(settings.AUTH_USER_MODEL, related_name="sentry_owner_set", null=True)
```
Do you actually need the related names? Or was this just from copy/paste?

iProgramStuff replied:
I thought they _could_ come in handy but I don't currently see a use case. Left them in "just in case""

iProgramStuff replied:
(Removed as per Slack)

mattrobenolt commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+
+        raise NotImplementedError("Unknown Owner")
+
+    def owner(self):
```
Copy/pasting this stuff over is fine, but make sure when you leverage using this model, try and leverage the Actor abstraction since this is exactly what it was designed for. Anywhere soemthing may be a User or Team, the Actor object should stand in for it.

iProgramStuff replied:
Gotcha. Right now I think search returns a User or a Team but it looks like I'll have to use a different parser for owner search.

billyvg committed:
feat(ui): Disable auto syncing of dark mode (#22159)

There has been a few complaints, so we are going to disable the auto syncing of dark mode while allowing manual toggle. I have intentionally left some code that will always evaluate to false, but will be added as a user option in the future.

billyvg committed:
build(gha): Run `api docs` test only on backend changes (#21828)

Refactored api docs js dependencies into its own `package.json` as well as making the tests skip if no api-docs related deps change (e.g. backend files + `api-docs/` dir)

Zylphrex committed:
feat(releases): Add slow lcp table to release details (#22141)

This change adds a new dropdown option to the release details transactions list
to sort by slowest LCP for that release.

Zylphrex committed:
feat(releases): Prioritize the failure count display (#22169)

MeredithAnya committed:
ref(stacktrace-link): Actionable error message (#22076)

* ref(stacktrace-link): Actionable error message

Zylphrex committed:
feat(discover): Add failure_count function to discover (#22175)

Opening the corresponding discover query from release details will add
failure_count as a column. However, because this function has yet to be exposed
in discover, the column edit model does not render it correctly.

wmak committed:
feat(discover): Add a compare_numeric_aggregate function (#22147)

- This is primarily so we can do compare(p75>threshold), but I'm trying
  to set this up so we can use it in the future as well
- This is an aggregate so that it can work on aggregates, which means that 
  it will be added to the HAVINIG clause properly and won't cause a group by

wmak committed:
feat(release-perf): duration charts (#22174)

- feat(release-perf): Adding charts based on duration
  - These charts will show the number of transactions over the thresholds
    we have, duration>300ms and lcp>4000ms
- feat(performance): Linking to new release charts
  - When on the vitals view link to chart with count of bad lcp
  - Otherwise link to the chart with counts of bad duration

davidenwang committed:
ref(ts): Convert TimePicker component (#22153)

priscilawebdev committed:
ref(yarn-lock): Update yarn lock (#22187)

markstory committed:
ref(dashboards) Remove old widget models from django (#22131)

Stop tracking these tables in django so that they can be dropped.

Zylphrex committed:
fix(releases): Sort by failing transactions by default (#22180)

Sorting by throughput does not surface interesting transactions, change the
default to sort by failure count.

Zylphrex committed:
fix(releases): Add tpm filter to sort orders that are prone to noise (#22179)

Sorts on p50() and p75(measurements.lcp) are prone to noise due to low counts.
Adding an filter on the throughput helps reduce the noice.

markstory commented on src/sentry/migrations/0130_groupownermodel.py:
```diff
@@ -0,0 +1,53 @@
+                ('id', sentry.db.models.fields.bounded.BoundedBigAutoField(primary_key=True, serialize=False)),
+                ('owner_type', models.PositiveSmallIntegerField()),
+                ('date_added', models.DateTimeField(default=django.utils.timezone.now)),
+                ('group', sentry.db.models.fields.foreignkey.FlexibleForeignKey(db_constraint=False, on_delete=django.db.models.deletion.CASCADE, to='sentry.Group', unique=True)),
```
You will need to update group deletions for this relation.

iProgramStuff replied:
Thanks! Good catch!

markstory commented on src/sentry/models/groupowner.py:
```diff
@@ -0,0 +1,54 @@
+    def save(self, *args, **kwargs):
+        assert not (self.user_id is not None and self.team_id is not None) and not (
+            self.user_id is None and self.team_id is None
+        ), "Must have Team or User, not both"
```
Something like 

```python
keys = list(filter(None, [self.user_id, self.team_id]))
assert len(keys) == 1, "Must have team or user, not both"
```

It is a bit shorter, and will grow more easily if you need it.

dashed committed:
ref(dashboards): Delete dashboards synchronously (#22167)

robinrendle committed:
ref(css): removed list group bootstrap styles (#22182)

billyvg committed:
test(acceptance): Fix flaky snapshots for performance and eventsv2 (#20920)

These tests have flaky snapshots due to the event not being available when the page is loaded. Change tests to look for a non empty state so that it will attempt to retry snapshot with correct content.

iProgramStuff committed:
Update src/sentry/models/groupowner.py

Co-authored-by: Matt Robenolt <matt@ydekproductions.com>

iProgramStuff committed:
Update src/sentry/models/groupowner.py

Co-authored-by: Matt Robenolt <matt@ydekproductions.com>

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/groupownermigration

iProgramStuff committed:
Updateing migration lockfile

iProgramStuff committed:
Removing related name..again

iProgramStuff committed:
Re-applying feedback

iProgramStuff committed:
Updating migration
